File: Sources/Chunk/MapRenderSampling.swift
import Foundation

struct MapChunkSampleAxis: Equatable {
  let startOffset: Int
  let endOffset: Int
  let representativeOffset: Int

  var span: Int { endOffset - startOffset + 1 }
}

/// Keeps the represented map extent unlimited while bounding the number of
/// LevelDB chunk decodes and temporary images required for one frame. When the
/// logical region becomes very large, one representative chunk is sampled for
/// each tile and stretched across that tile. Taps still read the exact target
/// chunk directly from LevelDB.
struct MapChunkSamplingPlan: Equatable {
  let logicalSideChunks: Int
  let stride: Int
  let xAxis: [MapChunkSampleAxis]
  let zAxis: [MapChunkSampleAxis]

  var sampleCount: Int {
    let (product, overflow) = xAxis.count.multipliedReportingOverflow(by: zAxis.count)
    return overflow ? Int.max : product
  }

  var isDownsampled: Bool { stride > 1 }

  static func make(
    sideChunks: Int,
    leftChunks: Int,
    maximumSamplesPerAxis: Int
  ) -> MapChunkSamplingPlan {
    let side = max(1, sideChunks)
    let left = min(max(0, leftChunks), side - 1)
    let right = side - left - 1
    let maximumSamples = max(1, maximumSamplesPerAxis)
    let stride = max(1, ceilingDivide(side, by: maximumSamples))
    let axis = makeAxis(start: -left, end: right, stride: stride)
    return MapChunkSamplingPlan(
      logicalSideChunks: side,
      stride: stride,
      xAxis: axis,
      zAxis: axis
    )
  }

  static func safeChunkCoordinate(center: Int32, offset: Int) -> Int32? {
    let (value, overflow) = Int64(center).addingReportingOverflow(Int64(offset))
    guard !overflow, value >= Int64(Int32.min), value <= Int64(Int32.max) else {
      return nil
    }
    return Int32(value)
  }

  private static func ceilingDivide(_ value: Int, by divisor: Int) -> Int {
    let quotient = value / divisor
    return quotient + (value % divisor == 0 ? 0 : 1)
  }

  private static func makeAxis(start: Int, end: Int, stride: Int) -> [MapChunkSampleAxis] {
    guard start <= end else { return [] }
    var result = [MapChunkSampleAxis]()
    result.reserveCapacity(ceilingDivide(end - start + 1, by: stride))
    var tileStart = start
    while tileStart <= end {
      let remaining = end - tileStart
      let tileEnd = tileStart + min(stride - 1, remaining)
      let representative: Int
      if tileStart <= 0, tileEnd >= 0 {
        representative = 0
      } else {
        representative = tileStart + (tileEnd - tileStart) / 2
      }
      result.append(
        MapChunkSampleAxis(
          startOffset: tileStart,
          endOffset: tileEnd,
          representativeOffset: representative
        ))
      guard tileEnd < end else { break }
      tileStart = tileEnd + 1
    }
    return result
  }
}

File: Windows/MCBEEditor.Core/Chunk/MapRenderSampling.cs
namespace MCBEEditor.Core.Chunk;

public readonly record struct MapSampleAxis(int StartOffset, int EndOffset, int RepresentativeOffset)
{
    public int Span => EndOffset - StartOffset + 1;
}

/// <summary>
/// Bounds the number of LevelDB chunk decodes while preserving the complete
/// represented world extent. Large logical regions are divided into tiles and
/// one representative chunk is decoded for each tile; the tile is then drawn
/// with nearest-neighbour scaling. Clicks can still resolve the exact world
/// coordinate independently of the representative sample.
/// </summary>
public sealed record MapRenderSamplingPlan(
    int LogicalSideChunks,
    int Stride,
    IReadOnlyList<MapSampleAxis> XAxis,
    IReadOnlyList<MapSampleAxis> ZAxis)
{
    public int SampleCount
    {
        get
        {
            var product = (long)XAxis.Count * ZAxis.Count;
            return product > int.MaxValue ? int.MaxValue : (int)product;
        }
    }

    public bool IsDownsampled => Stride > 1;

    public static MapRenderSamplingPlan Create(int sideChunks, int leftChunks, int maximumSamplesPerAxis)
    {
        var side = Math.Max(1, sideChunks);
        var left = Math.Clamp(leftChunks, 0, side - 1);
        var right = side - left - 1;
        var maximumSamples = Math.Max(1, maximumSamplesPerAxis);
        var stride = Math.Max(1, CeilingDivide(side, maximumSamples));
        var axis = MakeAxis(-left, right, stride);
        return new MapRenderSamplingPlan(side, stride, axis, axis);
    }

    public static bool TryChunkCoordinate(int center, int offset, out int coordinate)
    {
        var value = (long)center + offset;
        if (value is < int.MinValue or > int.MaxValue)
        {
            coordinate = 0;
            return false;
        }
        coordinate = (int)value;
        return true;
    }

    private static int CeilingDivide(int value, int divisor)
        => value / divisor + (value % divisor == 0 ? 0 : 1);

    private static IReadOnlyList<MapSampleAxis> MakeAxis(int start, int end, int stride)
    {
        if (start > end) return Array.Empty<MapSampleAxis>();
        var result = new List<MapSampleAxis>(CeilingDivide(end - start + 1, stride));
        var tileStart = start;
        while (tileStart <= end)
        {
            var tileEnd = tileStart + Math.Min(stride - 1, end - tileStart);
            var representative = tileStart <= 0 && tileEnd >= 0
                ? 0
                : tileStart + (tileEnd - tileStart) / 2;
            result.Add(new MapSampleAxis(tileStart, tileEnd, representative));
            if (tileEnd >= end) break;
            tileStart = tileEnd + 1;
        }
        return result;
    }
}
